onnxruntime/contrib_ops/cuda/bert/group_query_attention_helper.h

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/common/common.h"
+#include "core/providers/common.h"
+#include "contrib_ops/cpu/bert/attention_common.h"
+
+namespace onnxruntime {
+namespace contrib {
+namespace group_query_attention_helper {
+
+template <typename T>
+Status CheckInputs(const T* query,
+                   const T* key,
+                   const T* value,
+                   const T* past_key,
+                   const T* past_value,
+                   void* parameters,
+                   int num_heads,
+                   int kv_num_heads,
+                   const T* past_seq_len,
+                   bool is_past_bsnh,
+                   float scale) {
+  //     past_key                   : (B, S*, N_k, H) or (B, N_k, S*, H)
+  //     past_value                 : (B, S*, N_k, H) or (B, N_k, S*, H)
+  // no packing for q/k/v:
+  //     query            (Q)       : (B, S, D)
+  //     key              (K)       : (B, L, D_kv) or (B, N_k, S*, H)
+  //     value            (V)       : (B, L, D_kv) or (B, N_k, S*, H)
+
+  AttentionQkvFormat qkv_format = Q_K_V_BSNH;
+  AttentionQkvFormat past_kv_format = Q_K_V_BSNH;
+
+  const auto& query_dims = query->Shape().GetDims();
+  const auto& key_dims = key->Shape().GetDims();
+  const auto& value_dims = value->Shape().GetDims();
+
+  if (query_dims.size() != 3) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Input 'query' is expected to have 3 dimensions, got ",
+                           query_dims.size());
+  }
+
+  int batch_size = static_cast<int>(query_dims[0]);
+  int sequence_length = static_cast<int>(query_dims[1]);
+  int q_hidden_size = static_cast<int>(query_dims[2]);
+  int head_size = static_cast<int>(q_hidden_size) / num_heads;
+
+  int kv_sequence_length = sequence_length;
+  int kv_hidden_size = (key_dims.size() == 3)
+                           ? static_cast<int>(key_dims[2])
+                           : (kv_num_heads * static_cast<int>(key_dims[3]));
+
+  int max_sequence_length = 0;
+  if (past_key != nullptr && past_value != nullptr) {
```

Comment:
Suggest to add a function to check (B, N_k, S*, H) or ((B, S*, N_k, H) since we know B, N_k and H values, and return S* value.
Then we can reuse the function to check key/value and past_key/past_value when they are 4D